dds/DCPS/XTypes/DynamicData.cpp

```diff
@@ -2193,6 +2239,458 @@ const char* DynamicData::typekind_to_string(TypeKind tk) const
+    }
+    case XTypes::TK_UINT64: {
+      ACE_CDR::ULongLong my_ulonglong;
+      if (dd.get_uint64_value(my_ulonglong, XTypes::DISCRIMINATOR_ID) != DDS::RETCODE_OK) {
+        if (DCPS::log_level >= DCPS::LogLevel::Error) {
+          ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: print_dynamic_data: failed to get_uint64_value\n"));
+        }
+        return false;
+      }
+      type_string += " = " + DCPS::to_dds_string(my_ulonglong) + "\n";
+      break;
+    }
+    case XTypes::TK_BOOLEAN: {
+      ACE_CDR::Boolean my_bool;
+      if (dd.get_boolean_value(my_bool, XTypes::DISCRIMINATOR_ID) != DDS::RETCODE_OK) {
+        if (DCPS::log_level >= DCPS::LogLevel::Error) {
+          ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: print_dynamic_data: failed to get_boolean_value\n"));
+        }
+        return false;
+      }
+      type_string += " = " + DCPS::to_dds_string(my_bool) + "\n";
+      break;
+    }
+    case XTypes::TK_BYTE: {
+      ACE_CDR::Octet my_byte;
+      if (dd.get_byte_value(my_byte, XTypes::DISCRIMINATOR_ID) != DDS::RETCODE_OK) {
+        if (DCPS::log_level >= DCPS::LogLevel::Error) {
+          ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: print_dynamic_data: failed to get_byte_value\n"));
+        }
+        return false;
+      }
+      type_string += " = " + DCPS::to_dds_string(my_byte) + "\n";
+      break;
+    }
+    case XTypes::TK_CHAR8: {
+      ACE_CDR::Char my_char;
+      if (dd.get_char8_value(my_char, XTypes::DISCRIMINATOR_ID) != DDS::RETCODE_OK) {
+        if (DCPS::log_level >= DCPS::LogLevel::Error) {
+          ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: print_dynamic_data: failed to get_char8_value\n"));
+        }
+        return false;
+      }
+      type_string += DCPS::String(" = \'") + my_char + "\'\n";
+      break;
+    }
+    case XTypes::TK_CHAR16: {
+      ACE_CDR::WChar my_wchar;
+      if (dd.get_char16_value(my_wchar, XTypes::DISCRIMINATOR_ID) != DDS::RETCODE_OK) {
+        if (DCPS::log_level >= DCPS::LogLevel::Error) {
+          ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: print_dynamic_data: failed to get_char16_value\n"));
+        }
+        return false;
+      }
+      type_string += " = \'" + DCPS::to_dds_string(my_wchar) + "\'\n";
+      break;
+    }
+    }
+    if (item_count == 2) {
+      XTypes::DynamicTypeMember_rch temp_dtm;
+      XTypes::DynamicTypeMembersById dtmbi;
+      DCPS::String decoy_string;
+      dd.type()->get_all_members(dtmbi);
+      for (XTypes::DynamicTypeMembersById::iterator iter = dtmbi.begin(); iter != dtmbi.end(); ++iter) {
+        dd.type()->get_member(temp_dtm, iter->first);
+        if (dd.get_complex_value(dd, iter->first) == DDS::RETCODE_OK) {
+          if (print_dynamic_data(dd, decoy_string, indent)) {
+            member_name = temp_dtm->get_descriptor().name;
+            type_name = temp_dtm->get_descriptor().get_type()->get_descriptor().name;
+            type_string += indent + type_name + " " + member_name;
+            print_dynamic_data(dd, type_string, indent);
```

Comment:
This prints the selected member the second time, is it intended?

Reply:
Yes it is intended. The first call to `print_dynamic_data` is passed `decoy_string` which is a dummy to catch/suppress the errors that result when the non-selected branches are printed.  Once the selected branch is found, the if will be entered and `print_dynamic_data` will be called with `type_string`.

Reply:
`dd.get_complex_value(dd, iter->first)` returns `RETCODE_OK` if `iter->first` is the ID of the selected branch and returns `RETCODE_ERROR` in other cases. So the code inside the outer `if` block only executes for the selected branch.